src/libraries/System.Private.CoreLib/src/System/Text/Latin1Utility.cs

```diff
@@ -945,6 +945,13 @@ private static unsafe nuint NarrowUtf16ToLatin1_Sse2(char* pUtf16Buffer, byte* p
         /// </summary>
         public static unsafe void WidenLatin1ToUtf16(byte* pLatin1Buffer, char* pUtf16Buffer, nuint elementCount)
         {
+            if (((nuint)pUtf16Buffer & 1) != 0)
+            {
+                // Input isn't char aligned, we won't be able to vectorize.
+                WidenLatin1ToUtf16_MisalignedAddress(pLatin1Buffer, pUtf16Buffer, elementCount);
```

Comment:
I am not sure I understand the need of this, I am seeing that we already align data in https://github.com/dotnet/runtime/blob/4218f4c5cdcd863f7e094966a50c3c4f306bd4d6/src/libraries/System.Private.CoreLib/src/System/Text/Latin1Utility.cs#L990-L1015, don't we?

Reply:
If the address that `pUtf16Buffer` points to is naturally aligned, then this logic should work as expected.

The problem is:

```csharp
char* pCurrentWriteAddress = pUtf16Buffer + currentOffset;
```

which does not effectively ensure correct alignment for a misaligned `char*`.

Reply:
@EgorBo As an alternative approach, change the following to `Sse2.Store` and accept the performance hit pre-Nehalem.

https://github.com/dotnet/runtime/blob/4218f4c5cdcd863f7e094966a50c3c4f306bd4d6/src/libraries/System.Private.CoreLib/src/System/Text/Latin1Utility.cs#L899

Reply:
We should just use unaligned load/store here. It may have some small hit on hardware that's more than 10-12 years old, but it simplifies things overall and will be increasingly rare to encounter such hardware over time. Plus, we don't use aligned loads/stores in most of our workloads already, so this won't be any different.

Opportunistically aligning the data is still goodness and will avoid cache line splits.

Reply:
It looks like `Sse2.Store` approach is better since unaligned char* should be rare? (e.g. char[] is always aligned)

Reply:
The consideration would be that `Sse2.Store` doesn't allow containment on pre-AVX hardware and on pre-AVX hardware `movups` was frequently slower than `movaps` even if the data was definitely aligned.

But, given the age of the hardware involved and the logic we use elsewhere, it shouldn't be a huge consideration (and the cost also wasn't huge, typically an extra cycle or so; the split cache line accesses were much worse and closer to 20 cycles for loads).

Reply:
> It looks like `Sse2.Store` approach is better since unaligned char* should be rare? (e.g. char[] is always aligned)

A potential case is the public API `System.Text.Encoding.GetChars(System.Byte*, System.Int32, System.Char*, System.Int32)`.

Reply:
> WidenLatin1ToUtf16_Sse2 and NarrowUtf16ToLatin1_Sse2 are the only methods in the repository that use StoreAligned.

Right, and so preserving the alignment here won't make a big difference to downlevel hardware. We should just normalize to not using `StoreAligned` so the code can be simpler and robust in the face of unalignable data.